src/plopp/widgets/linecut_tool.py

```diff
@@ -0,0 +1,121 @@
+from .tools import DrawLinesTool
+
+from functools import partial
+import numpy as np
+import scipp as sc
+from scipy.interpolate import interp2d
+from typing import Any, Dict, Optional
+
+
+def interpolator(xy, interp_func, resolution, data_unit, distance_unit):
+    x1 = xy['x'][0]
+    y1 = xy['y'][0]
+    x2 = xy['x'][1]
+    y2 = xy['y'][1]
+
+    xvalues = np.linspace(x1, x2, resolution)
+    yvalues = np.linspace(y1, y2, resolution)
+
+    profile = np.array([interp_func(x, y)[0] for x, y in zip(xvalues, yvalues)])
+    out = sc.DataArray(data=sc.array(dims=['distance'], values=profile, unit=data_unit),
+                       coords={
+                           'distance':
+                           sc.array(dims=['distance'],
+                                    values=np.linalg.norm(
+                                        [xvalues - xvalues[0], yvalues - yvalues[0]],
+                                        axis=0),
+                                    unit=distance_unit)
+                       })
+    return out
+
+
+class LineCutTool(DrawLinesTool):
+    """
+    Tool to draw cuts/profiles of a 2d data array along a arbitrary lines.
+
+      - When a line is added to the 2d figure, make a new cut in the 1d figure.
+      - When a line or one of its vertices is dragged on the 2d figure, update the
+        corresponding cut in the 1d figure.
+      - When a line is removed from the 2d figure, remove the corresponding cut from
+        the 1d figure.
+
+    Parameters
+    ----------
+    figure:
+        The main (two_dimensional) figure where the tool will draw lines.
+    fig1d:
+        The secondary figure where the tool will display the cut results.
+    data:
+        If there are more than one artists (2d meshes) on the main figure, we need to
+        specify which artist should the cuts be taken from. Typically this is not the
+        case, the data can just be determined from the main figure and this argument can
+        be omitted.
+    resolution:
+        The number of sampling points to use along the line cut.
+    """
+
+    def __init__(self,
+                 figure: View,
+                 fig1d: View,
+                 data: Optional[sc.DataArray] = None,
+                 resolution: int = 100):
+        if not isinstance(figure, Figure2d):
+            raise TypeError("The line cut tool is only designed to work on 2d figures.")
+        if data is None:
+            if len(figure.graph_nodes) > 1:
+                raise ValueError("Cannot automatically generate tool because figure "
+                                 "has more than one artist.")
+            data = list(figure.graph_nodes.values())[0].request_data()
+        self._data_array = data
+        if (self._data_array.coords[figure.dims['x']].unit == self._data_array.coords[
+                figure.dims['y']].unit):
+            distance_unit = self._data_array.coords[figure.dims['x']].unit
+        else:
+            distance_unit = ''
```

Comment:
This is a bit weird. Maybe a unit mismatch would indicate that doing this is just conceptually meaningless anyway?